### PR TITLE
Linting updates

### LINKED ISSUE #5410: Simplify linting process
1. Eliminate HTMLHint and replace all rules with eslint ones
2. Eliminate custom indent/spacing functions - some parts already are or can be replaced by `vue/html-indent`, `vue/script-indent`, for some of them we'll need new eslint rules (mostly spacing stuff)
2. If possible, run prettier on the entire vue component, and using the eslint-plugin-prettier.
3. Run stylelint on the entire vue component after eslint/prettier has been run on it, to reduce callback/promise hell.

So `lint.js` would be simplified to:
1. If JS, run eslint with prettier plugin.
2. If CSS/SCSS/SASS/LESS, run stylelint (open question of whether we want to run prettier or not on our style files too).
3. If Vue component, run stylelint, then run eslint with prettier.

References: #5400

### User-facing consequences

None

### PR DESCRIPTION
**Pull request by rtibbles**
## Summary
* Upgrades all linting related packages in kolibri-tools
* Adds @rushstack package to ensure eslint module resolution from kolibri-tools (more details here: https://github.com/microsoft/rushstack/tree/main/eslint/eslint-patch#what-it-does)
* Creates eslint linting rules to replace our previous component line spacing rules (2 spaces between blocks, 1 empty space at the beginning and end of a populated block)
* Drops HTMLHint completely
* Creates eslint linting rules to replace our previous use of HTMLHint ([rules used](https://github.com/learningequality/kolibri/blob/90ec6518156a2648439b9592dd8b0fb3c5d5bff2/packages/kolibri-tools/.htmlhintrc.js)):
  * Ensuring `id` unique
  * Ensuring `img` `src` attribute is not empty
  * Enforcing `kebab-case` compatible `class` names
  * All other rules should be enforced by running prettier on the entire Single File Component
* Simplifies the `lint` function in `lint.js` by always running `prettier` on all files, running `eslint` on `js` and `vue` files, and then stylelint on style files and vue components.
* This PR deliberately does not implement prettier formatting with the `eslint-prettier-plugin`, as it reduces performance and can lead to more squiggly red lines in IDEs that will later be autofixed.
* Ensures consistent reporting by comparing source with formatted results and reporting an additional error if it is different.

Note: This does not lint any pre-existing files to limit the diff at the moment (and to prevent massive merge conflicts). 

## References
Fixes #5410
Supersedes #9660 and supersedes #9689 

## Reviewer guidance
It seems that the prettier upgrade does slightly change how [whitespace around functions](https://prettier.io/blog/2020/03/21/2.0.0.html#always-add-a-space-after-the-function-keyword-3903httpsgithubcomprettierprettierpull3903-by-j-f1httpsgithubcomj-f1-josephfrazierhttpsgithubcomjosephfrazier-sosukesuzukihttpsgithubcomsosukesuzuki-thorn0httpsgithubcomthorn0-7516httpsgithubcomprettierprettierpull7516-by-bakkothttpsgithubcombakkot) is handled which will cause a lot of changes to a lot of files.

For all [other breaking changes](https://prettier.io/blog/2020/03/21/2.0.0.html) that I could identify (and which had options) I set the option to induce the fewest changes.

@MisRob one last question - the latest eslint-vue integrates your [unused props](https://eslint.vuejs.org/rules/no-unused-properties.html#vue-no-unused-properties) PR (with support for methods, props, computed etc) - so this might be an opportunity to switch out for mainstream, but I did not attempt that as yet.

----

## Testing checklist

- [ ] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [ ] Critical and brittle code paths are covered by unit tests

## PR process

- [ ] PR has the correct target branch and milestone
- [ ] PR has 'needs review' or 'work-in-progress' label
- [ ] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

## Reviewer checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md

### CODE REVIEW

File: packages/eslint-plugin-kolibri/lib/rules/vue-component-block-padding.js
Comment (nucleogenesis) [inline packages/eslint-plugin-kolibri/lib/rules/vue-component-block-padding.js]: I think this type definition should live above the definition of `linebreak`

**Review by nucleogenesis — COMMENTED** [review]
Overall code looks good. I didn't manually test it - but if tests pass I suspect it should be all set.

File: packages/eslint-plugin-kolibri/lib/rules/vue-component-block-padding.js
Comment (rtibbles) [reply packages/eslint-plugin-kolibri/lib/rules/vue-component-block-padding.js]: Yes - it should.

Comment (rtibbles) [reply packages/eslint-plugin-kolibri/lib/rules/vue-component-block-padding.js]: Updated!

**Comment by MisRob** [comment]
Thank you, @rtibbles. Overall, it's looking good to me. The new lint file is much simpler than it used to be, I appreciate it (and your patience, this task looks like no small undertaking).

> @MisRob one last question - the latest eslint-vue integrates your [unused props](https://eslint.vuejs.org/rules/no-unused-properties.html#vue-no-unused-properties) PR (with support for methods, props, computed etc) - so this might be an opportunity to switch out for mainstream, but I did not attempt that as yet.

Sounds good. There are already lots of useful updates in this PR, so whatever works for you.

Also, I'm not sure if this helps, but noticed that [the issue that motivated `rushstack/eslint`](https://github.com/eslint/eslint/issues/3458) has some [recent updates from two weeks ago](https://github.com/eslint/eslint/issues/3458#issuecomment-1239811334):

> Hi folks! Just an update that the new ESLint configuration has now been enabled as an experiment. This new config system eliminates the special treatment of shared configs so that you can include plugins as direct dependencies. Read about it here:
>
> https://eslint.org/blog/2022/08/new-config-system-part-1/
> https://eslint.org/blog/2022/08/new-config-system-part-2/
> https://eslint.org/blog/2022/08/new-config-system-part-3/
>
> This is the long-term solution to the problem in this issue. We know it took a while, but this was a really complex problem that we needed to think through.
>
> Because we now have this solution implemented, I'm closing this issue. Please try out the new config system and leave your feedback in the discussions.

**Comment by MisRob** [comment]
I didn't get to read through the new rules but can see they all have test files so feeling confident

**Comment by rtibbles** [comment]
> Also, I'm not sure if this helps, but noticed that https://github.com/eslint/eslint/issues/3458 has some https://github.com/eslint/eslint/issues/3458#issuecomment-1239811334:

They closed it about a day or two after I last looked at it! Great, I'll update accordingly!

**Comment by rtibbles** [comment]
> They closed it about a day or two after I last looked at it! Great, I'll update accordingly!

I looked - I think it might take quite a bit more work to do this migration to their new flat configuration, so I think I'll leave this for now, and open a follow up issue, once this is merged.

File: kolibri/core/assets/src/views/sortable/DragContainer.vue
Comment (bjester) [inline kolibri/core/assets/src/views/sortable/DragContainer.vue]: This is a confusing change? This line but on the line below? seems inconsistent

File: kolibri/core/assets/src/views/sync/SelectDeviceModalGroup/DeviceConnectingModal.vue
Comment (bjester) [inline kolibri/core/assets/src/views/sync/SelectDeviceModalGroup/DeviceConnectingModal.vue]: Not sure what prettier / linting rule is forcing this but it seems not ideal. I don't see any reason why the cases inside a switch should not be indented similarly to how you indent something in a function or `if` block

File: kolibri/core/assets/src/views/sync/SelectDeviceModalGroup/DeviceConnectingModal.vue
Comment (bjester) [inline kolibri/core/assets/src/views/sync/SelectDeviceModalGroup/DeviceConnectingModal.vue]: Hmm, is this even supposed to be here?

File: kolibri/plugins/coach/assets/src/views/common/HeaderWithOptions.vue
Comment (bjester) [inline kolibri/plugins/coach/assets/src/views/common/HeaderWithOptions.vue]: Another questionable discrepancy of indented comments

**Review by bjester — COMMENTED** [review]
There are some questionable indentation changes with comments, particularly inconsistent, and `switch` blocks that do not make much sense

File: kolibri/core/assets/src/views/sortable/DragContainer.vue
Comment (rtibbles) [reply kolibri/core/assets/src/views/sortable/DragContainer.vue]: Interesting - I caught the formatter doing something like this for the style block, but had already setup the vue script indent rule, but seems not to be working for comments specifically?

**Comment by rtibbles** [comment]
Have fixed what seemed to be the underlying causes of the comment and switch case indentations.

I think this is mostly ready to go, but I'd like to hold off on a final merge with one last rebase after the 0.16 into develop merge PR is merged (we may need to do two of these, as there is one open issue for 0.16.2 that might need a frontend tweak).

**Comment by rtibbles** [comment]
Filed a follow up issue for eslint/vue updates here: https://github.com/learningequality/kolibri/issues/12199

**Comment by marcellamaki** [comment]
let's go ahead with this this week as we are planning to move to our new branches approach, I'll approve once the merge conflicts are resolved

**Review by marcellamaki — APPROVED** [review]

**Review by marcellamaki — APPROVED** [review]

**Pull request merged**
